middle_end/flambda2/types/flambda2_types.ml

```diff
@@ -64,3 +64,63 @@ let join ?bound_name central_env ~left_env ~left_ty ~right_env ~right_ty =
   match join ?bound_name join_env left_ty right_ty with
   | Unknown -> unknown_like left_ty
   | Known ty -> ty
+
+let extract_symbol_approx env symbol find_code =
+  let rec type_to_approx (ty : t) : _ Value_approximation.t =
+    let expanded = Expand_head.expand_head env ty in
+    match Expand_head.Expanded_type.descr expanded with
+    | Unknown | Bottom -> Value_unknown
+    | Ok
+        ( Naked_immediate _ | Naked_float _ | Naked_int32 _ | Naked_int64 _
+        | Naked_nativeint _ | Rec_info _ | Region _ ) ->
+      Misc.fatal_error
+        "Typing_env.Serializable.to_closure_conversion_approx: Wrong kind"
+    | Ok (Value ty) -> begin
+      match ty with
+      | Array _ | String _ | Boxed_float _ | Boxed_int32 _ | Boxed_int64 _
+      | Boxed_nativeint _ | Mutable_block _ ->
+        Value_unknown
+      | Closures { by_closure_id; alloc_mode = _ } -> (
+        match Row_like_for_closures.get_singleton by_closure_id with
+        | None -> Value_unknown
+        | Some ((closure_id, _contents), closures_entry) -> begin
+          match Closures_entry.find_function_type closures_entry closure_id with
+          | Bottom | Unknown -> Value_unknown
+          | Ok function_type ->
+            let code_id = Function_type.code_id function_type in
+            let code = find_code code_id in
+            (* CR vlaviron: Should we fail if [code] is [None] ? *)
```

Comment:
The answer is no, it can happen if the function isn't inlinable for instance. in this case it's still useful to have the approximation, to be able to generate direct calls. I'll remove the comment.

Reply:
Done.